Refactor legacy `unknwon/com` package, improve golangci lint

wxiaoguang opened the pull request:
In this PR, the following refactoring are done. The main purpose is to refactor the legacy `unknwon/com`.

1. Remove most imports of `unknwon/com`, now only `util/legacy.go` imports the legacy `unknwon/com`, preparing for next round refactoring
2. Use golangci's depguard to process denied packages, we do not need to maintain our gitea-vet deny-list anymore in future.
3. Fix some incorrect values in golangci.yml, for example, the version should be quoted string `"1.18"`, instead of a float number `1.18`
4. Use correctly escaped content for `go-import` and `go-source` meta tags
5. Refactor `com.Expand` to our stable (and the same fast) `vars.Expand`. 
    * The `vars.Expand` can detect errors and show them in logs or to end users. 
    * To keep good user experience, our `vars.Expand` can still return partially rendered content even if the template is not good (eg: key mistach).
    * The reason is: in some cases (eg: merge message processing, or generating readme for repo init), we should show the partially rendered templates to users instead of frighten them by 500 errors again and again
    * Keep the same behavior and result as the legacy `com.Expand`

Gusted commented on modules/context/csrf.go:
```diff
@@ -162,7 +163,12 @@ func prepareOptions(options []CsrfOptions) CsrfOptions {
 	// Defaults.
 	if len(opt.Secret) == 0 {
-		opt.Secret = string(com.RandomCreateBytes(10))
+		randBytes, err := util.CryptoRandomBytes(8)
```
Just to note for other reviewers/maintainers that the reduction of this isn't a concern. 

Old situation:
The `com.RandomCreateBytes` actually doesn't give you the full 8 bits of randomness per byte(so 80 bits in the old case), because it reduces itself to only use the alphanum: https://github.com/unknwon/com/blob/b41c64acd94be7e673c9c8301344d31cce99e06c/string.go#L130 which only gives 6 bits per byte of randomness so it would only receive 48 bits of randomness.

New situation:
With `CryptoRandomBytes` the full 64 bits would be random.

wxiaoguang replied:
And before, IIRC the `com.RandomCreateBytes` uses a math PRNG ..... not crypto-safe.

Gusted commented on modules/markup/html.go:
```diff
@@ -838,7 +838,14 @@ func issueIndexPatternProcessor(ctx *RenderContext, node *html.Node) {
+			res, err := vars.Expand(ctx.Metas["format"], ctx.Metas)
+			if err != nil {
+				// here we could just log the error and continue the rendering
+				log.Error("unable to expand template vars for ref %s, err:%s", ref.Issue, err.Error())
```
```suggestion
				log.Error("unable to expand template vars for ref %s, err: %v", ref.Issue, err)
```

Gusted commented on modules/repository/init.go:
```diff
@@ -250,8 +249,13 @@ func prepareRepoCommit(ctx context.Context, repo *repo_model.Repository, tmpDir,
+	res, err := vars.Expand(string(data), match)
+	if err != nil {
+		// here we could just log the error and continue the rendering
+		log.Error("unable to expand template vars for repo README: %s, err: %s", opts.Readme, err.Error())
```
```suggestion
		log.Error("unable to expand template vars for repo README: %s, err: %v", opts.Readme, err)
```

Gusted commented on modules/util/path.go:
```diff
@@ -173,3 +177,22 @@ func FileURLToPath(u *url.URL) (string, error) {
+
+// HomeDir returns path of '~'(in Linux) on Windows,
+// it returns error when the variable does not exist.
+func HomeDir() (home string, err error) {
+	if isOSWindows() {
+		home = os.Getenv("USERPROFILE")
+		if home == "" {
+			home = os.Getenv("HOMEDRIVE") + os.Getenv("HOMEPATH")
+		}
+	} else {
+		home = os.Getenv("HOME")
+	}
+
+	if home == "" {
+		return "", errors.New("cannot get home directory")
+	}
+
+	return home, nil
```
Could we use something along the lines of:

```go
    user, err := user.Current()
    if err != nil {
        return err
    }
    return user.HomeDir
```

wxiaoguang replied:
Not sure, it's breaking. I believe some user run Gitea with "HOME=xxx" and mismatched uid ......

wxiaoguang replied:
A new comment added:
```
	// TODO: some users run Gitea with mismatched uid  and "HOME=xxx" (they set HOME=xxx by environment manually)
	// so at the moment we can not use `user.Current().HomeDir`
```

Gusted commented on modules/util/util.go:
```diff
@@ -181,3 +183,23 @@ func ToUpperASCII(s string) string {
+}
+
+// AESGCMDecrypt (from unknwon/com): decrypts ciphertext with the given key using AES in GCM mode. should be replaced.
+func AESGCMDecrypt(key, ciphertext []byte) ([]byte, error) {
```
Ditto

Gusted commented on modules/util/util.go:
```diff
@@ -181,3 +183,23 @@ func ToUpperASCII(s string) string {
+
+// AESGCMDecrypt (from unknwon/com): decrypts ciphertext with the given key using AES in GCM mode. should be replaced.
+func AESGCMDecrypt(key, ciphertext []byte) ([]byte, error) {
+	return com.AESGCMDecrypt(key, ciphertext)
```
```suggestion
		block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	size := gcm.NonceSize()
	if len(ciphertext)-size <= 0 {
		return nil, errors.New("ciphertext is empty")
	}

	nonce := ciphertext[:size]
	ciphertext = ciphertext[size:]

	plainText, err := gcm.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return nil, err
	}

	return plainText, nil
```

Gusted reviewed (commented):
Not sure why we're changing from `%v", err` -> `%s", err.Error()`

6543 commented on modules/util/util.go:
```diff
@@ -181,3 +183,62 @@ func ToUpperASCII(s string) string {
+	return append(nonce, ciphertext...), nil
+}
+
+// AESGCMDecrypt (from unknwon/com): decrypts ciphertext with the given key using AES in GCM mode. should be replaced.
```
```suggestion
// AESGCMDecrypt decrypts ciphertext with the given key using AES in GCM mode.
```

6543 commented on modules/util/util.go:
```diff
@@ -181,3 +183,62 @@ func ToUpperASCII(s string) string {
+	return com.ToSnakeCase(str)
+}
+
+// AESGCMEncrypt (from unknwon/com): encrypts plaintext with the given key using AES in GCM mode. should be replaced.
```
```suggestion
// AESGCMEncrypt: encrypts plaintext with the given key using AES in GCM mode.
```

6543 commented on modules/util/copy.go:
```diff
@@ -5,7 +5,7 @@
+	"github.com/unknwon/com" //nolint:depguard
 )
 
 // CopyFile copies file from source to target path.
```
If i remember correctly golang v1.16 did add some features to std lib

wxiaoguang replied:
If we can find a replacement, please edit on this PR directly, at the moment I can't 😂

wxiaoguang commented on modules/templates/vars/vars.go:
```diff
@@ -0,0 +1,109 @@
+
+// Expand replaces all variables like {var} to match, if error occurs,
+// the error part doesn't change and is returned as it is.
+func Expand(template string, match map[string]string) (string, error) {
```
Do you mean `text/template`? I think there are different, I can not remember if there is a exactly the same function in std library.

wxiaoguang commented:
TBH, I wasn't going to rewrite the AESGCMEncrypt, etc.

I feel that most of us like to do the rewrite, then I can improve this PR with more changes and tests.

Will be back soon.

wxiaoguang commented:
Updated. Since the `AESGCM` related functions are updated by `Apply suggestions from code review`, I added necessary tests for them.

And now there is only one `util/legacy.go` imports `github.com/unknwon/com`.

I think this PR is big enough and gets enough changes. More refactoring to the `ToStr`/`ToSnakeCase` could be done in a separate PR with more tests.

lunny commented on modules/context/repo.go:
```diff
@@ -308,11 +308,17 @@ func EarlyResponseForGoGetMeta(ctx *Context) {
 		return
 	}
-	ctx.PlainText(http.StatusOK, com.Expand(`<meta name="go-import" content="{GoGetImport} git {CloneLink}">`,
+	res, err := vars.Expand(`<meta name="go-import" content="{GoGetImport} git {CloneLink}">`,
```
We can just use `fmt.Sprintf` to replace them.

wxiaoguang replied:
The code was incorrect before, because it's not escaped correctly. A simple Sprintf is also incorrect.

If you want this change, then I can update the code with correctly escaped content.

https://github.com/golang/go/blob/2ebe77a2fda1ee9ff6fd9a3e08933ad1ebaea039/src/cmd/go/internal/vcs/discovery.go#L54

wxiaoguang replied:
See the latest commit

wxiaoguang commented:
After these refactorings, now there are only 2 unmaintained packages:

* `unknwon/i18n`: pretty simple (200 lines) and I think it should be improved.
* `unknwon/paginater`: not complex (180 lines + 300 lines test), no obvious defects.

Do you think we should continue to refactor these 2 packages? If vote yes (up+1), I could do it if I have time.
